magicblock-committor-service/src/intent_executor/single_stage_executor.rs

```diff
@@ -0,0 +1,190 @@
+                    junk.push(transaction_strategy);
+                    return Ok(ExecutionOutput::SingleStage(value));
+                }
+                Err(err) => err,
+            };
+
+            // Attempt patching
+            let flow = self
+                .patch_strategy(
+                    &execution_err,
+                    &mut transaction_strategy,
+                    &base_intent,
+                )
+                .await?;
+            let cleanup = match flow {
+                ControlFlow::Continue(cleanup) => {
+                    info!(
+                        "Patched intent: {}. patched error: {:?}",
+                        base_intent.id, execution_err
+                    );
+                    cleanup
+                }
+                ControlFlow::Break(()) => {
+                    error!("Could not patch failed intent: {}", base_intent.id);
+                    break (execution_err, transaction_strategy);
+                }
+            };
+
+            if i >= RECURSION_CEILING {
+                error!(
+                    "CRITICAL! Recursion ceiling reached in intent execution."
+                );
+                break (execution_err, cleanup);
+            } else {
+                junk.push(cleanup);
+            }
+        };
+
+        // Special case
+        let committed_pubkeys = base_intent.get_committed_pubkeys();
+        if i < RECURSION_CEILING
+            && matches!(
+                execution_err,
+                TransactionStrategyExecutionError::CpiLimitError
+            )
+            && committed_pubkeys.is_some()
+        {
+            // With actions, we can't predict num of CPIs
+            // If we get here we will try to switch from Single stage to Two Stage commit
+            // Note that this not necessarily will pass at the end due to the same reason
+
+            // SAFETY: is_some() checked prior
+            let committed_pubkeys = committed_pubkeys.unwrap();
+            let (commit_strategy, finalize_strategy, cleanup) =
+                self.handle_cpi_limit_error(last_transaction_strategy);
+            junk.push(cleanup);
+            self.two_stage_execution_flow(
+                &committed_pubkeys,
+                commit_strategy,
+                finalize_strategy,
+                persister,
+            )
+            .await
+        } else {
+            junk.push(last_transaction_strategy);
+            let err = IntentExecutorError::from_strategy_execution_error(
+                execution_err,
+                |internal_err| {
+                    let signature = internal_err.signature();
+                    IntentExecutorError::FailedToFinalizeError {
+                        err: internal_err,
+                        commit_signature: signature,
+                        finalize_signature: signature,
+                    }
+                },
+            );
+
+            Err(err)
+        }
+    }
+
+    pub async fn patch_strategy(
```

Comment:
I'm not sure how/what we're patching here.
Could you please add a comment to method that explains that?